server/src/main/java/org/elasticsearch/cluster/coordination/CoordinationDiagnosticsService.java

```diff
@@ -819,7 +840,7 @@ private Scheduler.Cancellable fetchClusterFormationInfo(
     }
 
     void beginPollingRemoteMasterStabilityDiagnostic() {
-        assert ThreadPool.assertCurrentThreadPool(ClusterApplierService.CLUSTER_UPDATE_THREAD_NAME);
+        // Note that this method must be called from the system context because it calls internal transport actions
```

Comment:
could this be coded as an assertion on the thread context? ie. `ThreadContext#isSystemContext`

Reply:
It's a little more complicated than that because our test code doesn't run in the system context. I'll add a method to ThreadPool to work with unit tests like we do for assertCurrentThreadPool.